src/nupic/research/temporal_memory.py

```diff
@@ -23,16 +23,109 @@
 Temporal Memory implementation in Python.
 """
 
-from collections import defaultdict, namedtuple
+from collections import defaultdict
 from operator import mul
 
 from nupic.bindings.math import Random
 from nupic.research.connections import Connections
 
-
+from sys import maxint as MAX_INT
+from bisect import bisect_left
 
 EPSILON = 0.000001
 
+def excitedColumnsGenerator(activeColumns,
+                            activeSegments,
+                            matchingSegments,
+                            cellsPerColumn,
+                            connections):
+  """
+  Generator used for iterating over the lists of active columns,
+  active segments, and matching segments, each sorted by the column they
+  correspond to.
+
+  @param activeColumns    (list)   List of currently active columns
+  @param activeSegments   (list)   List of segments active from lateral input
+  @param matchingSegments (list)   List of segments matching from lateral input
+  @param cellsPerColumn   (int)    Number of cells per column in the tm
+  @param connections      (Object) Connections object of the tm
+
+  @return (dict){
+                  `column`               (int),
+                  `isActiveColumn`       (bool),
+                  `activeSegmentsBegin`  (int),
+                  `activeSegmentsEnd`    (int),
+                  `matchingSegmentsBegin`(int),
+                  `matchingSegmentsEnd`  (int)
```

Comment:
Rather than Begin and End values, maybe it would be better to use an `xrange`? This would continue the trend of "Make our Python code more readable than our C++ by using generators instead of iterators". In other words, this would return a dict with 'activeSegments' and 'matchingSegments' set to an xrange.